assets/js/blocks/product-collection/inspector-controls/taxonomy-controls.tsx

```diff
@@ -0,0 +1,265 @@
+
+	const debouncedSearch = useDebounce( setSearch, 250 );
+
+	const { searchResults, searchHasResolved }: SearchResultsState = useSelect(
+		( select ) => {
+			if ( ! search ) {
+				return { searchResults: EMPTY_ARRAY, searchHasResolved: true };
+			}
+
+			const { getEntityRecords, hasFinishedResolution } =
+				select( coreStore );
+
+			const selectorArgs = [
+				'taxonomy',
+				taxonomy.slug,
+				{
+					...BASE_QUERY,
+					search,
+					orderby: 'name',
+					exclude: termIds,
+					per_page: 20,
+				},
+			];
+			return {
+				searchResults: getEntityRecords( ...selectorArgs ),
+				searchHasResolved: hasFinishedResolution(
+					'getEntityRecords',
+					selectorArgs
+				),
+			};
+		},
+		[ search, termIds ]
+	);
+
+	// `existingTerms` are the ones fetched from the API and their type is `{ id: number; name: string }`.
+	// They are used to extract the terms' names to populate the `FormTokenField` properly
+	// and to sanitize the provided `termIds`, by setting only the ones that exist.
+	const existingTerms: Term[] = useSelect(
+		( select ) => {
+			if ( ! termIds?.length ) return EMPTY_ARRAY;
+			const { getEntityRecords } = select( coreStore );
+			return getEntityRecords( 'taxonomy', taxonomy.slug, {
+				...BASE_QUERY,
+				include: termIds,
+				per_page: termIds.length,
+			} );
+		},
+		[ termIds ]
+	);
+
+	// Update the `value` state only after the selectors are resolved
+	// to avoid emptying the input when we're changing terms.
+	useEffect( () => {
+		if ( ! termIds?.length ) {
+			setValue( EMPTY_ARRAY );
+		}
+
+		if ( ! existingTerms?.length ) return;
+
+		// Returns only the existing entity ids. This prevents the component
+		// from crashing in the editor, when non existing ids are provided.
+		const sanitizedValue = termIds.reduce(
+			(
+				accumulator: {
+					id: number;
+					value: string;
+				}[],
+				id
+			) => {
+				const entity = existingTerms.find( ( term ) => term.id === id );
+				if ( entity ) {
+					accumulator.push( {
+						id,
+						value: entity.name,
+					} );
+				}
+				return accumulator;
+			},
+			[]
+		);
+		setValue( sanitizedValue );
+	}, [ termIds, existingTerms ] );
+
+	// Update suggestions only when the query has resolved.
+	useEffect( () => {
+		if ( ! searchHasResolved ) return;
+		setSuggestions( searchResults.map( ( result ) => result.name ) );
+	}, [ searchResults, searchHasResolved ] );
+
+	const onTermsChange = ( newTermValues: FormTokenField.Value[] ) => {
+		const newTermIds = new Set< number >();
+		for ( const termValue of newTermValues ) {
+			const termId = getTermIdByTermValue(
+				searchResults,
+				termValue as string | Term
+			);
+			if ( termId ) {
+				newTermIds.add( termId );
+			}
+		}
+		setSuggestions( EMPTY_ARRAY );
+		onChange( Array.from( newTermIds ) );
+	};
+
+	return (
+		<div className="wc-block-editor-product-collection-inspector__taxonomy-control">
+			<FormTokenField
+				label={ taxonomy.name }
+				value={ value }
```

Comment:
I'm a bit confused by this, but if I'm not wrong, `value` can be directly inferred from `existingTerms`, right? In that case, we shouldn't need to store `value` in the state. Ie, we could do something like this:

```suggestion
				value={ ( existingTerms || [] ).map( ( { id, name } ) => ( {
					id,
					value: name,
				} ) ) }
```

and then get rid of the `state` value and the `useEffect()` associated with it. (I might have missed something, though :sweat_smile:)

Reply:
The reason for using useState is that when a user adds a new term, `existingTerms` becomes empty. Therefore, if we were to use `existingTerms` as you suggested, the input would become empty for a few seconds while the `existingTerms` are being fetched. However, I will take a look at this and see if any improvements can be made. Thanks for pointing this out 🙌🏻

Reply:
Oooh, you are right, I see what you mean. I wonder if we could use [`usePrevious()`](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/base/hooks/use-previous.ts), however. IMO it seems like a better hook for this use case. So we would use `existingTerms` by default, but if it's empty, we would use `previousExistingTerms` to make sure the input field doesn't become empty. What do you think? :thinking:

Reply:
Actually, I did consider using `usePrevious()`, but the implementation would be somewhat complex. For instance, if there is one term selected and the user select another term, the value of `existingTerms` would look like this:

<img width="218" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/5d083ff0-8626-4400-8454-fffb04d9869c">

As you can see in the console.log output above. There are multiple null and empty values before the second term is added to existing terms. This complicates determining when we should use the `usePrevious` value. Do you see the point? However, I will experiment with it today to see if I can figure out how to use `usePrevious`, which may require some code refactoring.

Reply:
Just for future reference: heads-up that `usePrevious()` accepts a [`validation` function](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/base/hooks/use-previous.ts#L18) so you can determine whether it updates or not. So in this case, it should be possible to make it so it doesn't update in certain circumstances.